extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authentication/db/cache/CoordinatorBasicAuthenticatorCacheNotifier.java

```diff
@@ -0,0 +1,280 @@
+import io.druid.discovery.DruidNodeDiscoveryProvider;
+import io.druid.guice.ManageLifecycle;
+import io.druid.guice.annotations.EscalatedClient;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.StringUtils;
+import io.druid.java.util.common.concurrent.Execs;
+import io.druid.java.util.common.lifecycle.LifecycleStart;
+import io.druid.security.basic.authentication.BasicHTTPAuthenticator;
+import io.druid.security.basic.BasicAuthDBConfig;
+import io.druid.server.DruidNode;
+import io.druid.server.security.Authenticator;
+import io.druid.server.security.AuthenticatorMapper;
+import org.jboss.netty.handler.codec.http.HttpChunk;
+import org.jboss.netty.handler.codec.http.HttpMethod;
+import org.jboss.netty.handler.codec.http.HttpResponse;
+import org.joda.time.Duration;
+
+import javax.ws.rs.core.MediaType;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+@ManageLifecycle
+public class CoordinatorBasicAuthenticatorCacheNotifier implements BasicAuthenticatorCacheNotifier
+{
+  private static final EmittingLogger LOG = new EmittingLogger(CoordinatorBasicAuthenticatorCacheNotifier.class);
+
+  private static final List<String> NODE_TYPES = Arrays.asList(
+      DruidNodeDiscoveryProvider.NODE_TYPE_BROKER,
+      DruidNodeDiscoveryProvider.NODE_TYPE_OVERLORD,
+      DruidNodeDiscoveryProvider.NODE_TYPE_HISTORICAL,
+      DruidNodeDiscoveryProvider.NODE_TYPE_PEON,
+      DruidNodeDiscoveryProvider.NODE_TYPE_ROUTER,
+      DruidNodeDiscoveryProvider.NODE_TYPE_MM
+  );
+
+  private final DruidNodeDiscoveryProvider discoveryProvider;
+  private final HttpClient httpClient;
+  private final Set<String> authenticatorsToUpdate;
+  private final Map<String, byte[]> serializedMaps;
+  private final Map<String, BasicAuthDBConfig> authenticatorConfigMap;
+  private final LifecycleLock lifecycleLock = new LifecycleLock();
+
+  private Thread notifierThread;
+
+  @Inject
+  public CoordinatorBasicAuthenticatorCacheNotifier(
+      AuthenticatorMapper authenticatorMapper,
+      DruidNodeDiscoveryProvider discoveryProvider,
+      @EscalatedClient HttpClient httpClient
+  )
+  {
+    this.discoveryProvider = discoveryProvider;
+    this.httpClient = httpClient;
+    this.authenticatorsToUpdate = new HashSet<>();
+    this.serializedMaps = new HashMap<>();
+    this.authenticatorConfigMap = new HashMap<>();
+    initAuthenticatorConfigMap(authenticatorMapper);
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    if (!lifecycleLock.canStart()) {
+      throw new ISE("can't start.");
+    }
+
+    try {
+      notifierThread = Execs.makeThread(
+          "CoordinatorBasicAuthenticatorCacheNotifier-notifierThread",
+          () -> {
+            while (!Thread.interrupted()) {
```

Comment:
I wonder why you created a single thread which repeatedly sends update notifications instead of using an ExecutorService and submitting a new task on addUpdate(). I guess this will be simpler.

Reply:
Hm, there wasn't a particular reason, do you see an advantage to using ExecutorService with new tasks?

Reply:
I think using ExecutorService and BlockingQueue will be simpler to implement/maintain the code. Please see the comments (https://github.com/druid-io/druid/pull/5099#discussion_r156345477 and https://github.com/druid-io/druid/pull/5099#discussion_r156347623)